edge-util/src/Microsoft.Azure.Devices.Edge.Util/SharedAccessSignature.cs

```diff
@@ -0,0 +1,227 @@
+            {
+                throw new UnauthorizedAccessException("The specified SAS token has expired.");
+            }
+
+            if (sasAuthorizationRule.PrimaryKey != null)
+            {
+                string primaryKeyComputedSignature = ComputeSignature(Convert.FromBase64String(sasAuthorizationRule.PrimaryKey), this.encodedAudience, this.expiry);
+
+                if (CryptographicOperations.FixedTimeEquals(Encoding.UTF8.GetBytes(this.Signature), Encoding.UTF8.GetBytes(primaryKeyComputedSignature)))
+                {
+                    return;
+                }
+            }
+
+            if (sasAuthorizationRule.SecondaryKey != null)
+            {
+                string secondaryKeyComputedSignature = ComputeSignature(Convert.FromBase64String(sasAuthorizationRule.SecondaryKey), this.encodedAudience, this.expiry);
+
+                if (CryptographicOperations.FixedTimeEquals(Encoding.UTF8.GetBytes(this.Signature), Encoding.UTF8.GetBytes(secondaryKeyComputedSignature)))
+                {
+                    return;
+                }
+            }
+
+            throw new UnauthorizedAccessException("The specified SAS token has an invalid signature. It does not match either the primary or secondary key.");
+        }
+
+        /// <summary>
+        /// Compute the signature string using the SAS fields.
+        /// </summary>
+        /// <param name="key">Key used for computing the signature.</param>
+        /// <returns>The string representation of the signature.</returns>
+        public static string ComputeSignature(byte[] key, string encodedAudience, string expiry)
+        {
+            var fields = new List<string>
+            {
+                encodedAudience,
+                expiry,
+            };
+
+            using var hmac = new HMACSHA256(key);
+            string value = string.Join("\n", fields);
+            return Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(value)));
+        }
+
+        private static IDictionary<string, string> ExtractFieldValues(string sharedAccessSignature)
+        {
+            string[] lines = sharedAccessSignature.Split();
+
+            if (!StringComparer.Ordinal.Equals(
+                    lines[0].Trim(),
+                    SharedAccessSignatureConstants.SharedAccessSignature)
+                || lines.Length != 2)
+            {
+                throw new FormatException("Malformed signature");
+            }
+
+            IDictionary<string, string> parsedFields = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            string[] fields = lines[1].Trim().Split(new string[] { SharedAccessSignatureConstants.PairSeparator }, StringSplitOptions.None);
+
+            foreach (string field in fields)
+            {
+                if (!string.IsNullOrEmpty(field))
+                {
+                    string[] fieldParts = field.Split(SharedAccessSignatureConstants.KeyValueSeparator, 2);
```

Comment:
This is different to the original: https://github.com/Azure/azure-iot-sdk-csharp/blob/feb2c4b6a6da326c0e9dc48d3f7c474ff97423ae/iothub/device/src/Common/Security/SharedAccessSignature.cs

It is also on purpose, as I found original would split on 64 encodings that had '=' in them

Reply:
I think I was wrong about this, I am reverting to what they have again. I thought this was necessary as I didn't see where we were handling encoding of the signature. But because the expected token being passed in is UrlEncoded (at least the signature part) there is no way for it to accidentally split more than it should!